### PR TITLE
BSP-Sbt ext-pt impl with persistent storage

### PR DESCRIPTION
**Pull request by AnthonyGrod**

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -700,8 +700,10 @@ lazy val javaDecompilerIntegration =
 lazy val intellijBspIntegration =
   newProject("intellij-bsp", file("scala/integration/intellij-bsp"))
     .dependsOn(scalaImpl)
+    .dependsOn(sbtImpl)
     .settings(
-      intellijPlugins += "org.jetbrains.bsp::nightly".toPlugin
+      intellijPlugins += "org.jetbrains.bsp::daily".toPlugin.withFallbackDownloadUrl("https://students.mimuw.edu.pl/~ag438477/intellij-bsp-2024.1.0-EAP.zip"),
```
Comment (agluszak) [inline build.sbt]: haha, that's very clever x)

File: scala/integration/intellij-bsp/src/org/jetbrains/plugins/scala/project/bsp/sbt/module/SbtBspModuleSettingsForm.java [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+
+import javax.swing.*;
+import javax.swing.border.EmptyBorder;
+import javax.swing.table.TableColumnModel;
+import java.awt.*;
+import java.lang.reflect.Method;
+import java.util.Collections;
+import java.util.ResourceBundle;
+
+@SuppressWarnings("unchecked")
```
Comment (agluszak) [inline scala/integration/intellij-bsp/src/org/jetbrains/plugins/scala/project/bsp/sbt/module/SbtBspModuleSettingsForm.java:22]: could you move it closer to the place where this warning actually happens?

**Review by agluszak — COMMENTED** [review]
lgtm code-wise, I will test how it works tomorrow

**Comment by agluszak** [comment]
```
[warn] Plugin not found at https://plugins.jetbrains.com/pluginManager?action=download&noStatistic=true&id=org.jetbrains.bsp&channel=daily&build=IC-241.14494.277
[warn] Trying to download from a fallback url https://students.mimuw.edu.pl/~ag438477/intellij-bsp-2024.1.0-EAP.zip
[error] org.jetbrains.sbtidea.download.FileDownloader$DownloadException: Not Found (404): https://students.mimuw.edu.pl/~ag438477/intellij-bsp-2024.1.0-EAP.zip
```

**Comment by AnthonyGrod** [comment]
Yes, I changed the name of the zip for an hour but now it's back and should work. Sorry @agluszak

**Comment by agluszak** [comment]
Nope, still 404

**Comment by AnthonyGrod** [comment]
Sorry, I had a typo. Could you try it now?

**Comment by agluszak** [comment]
Now sync works, but I'm getting warnings:

```
[warn] ### Installed build number is different form the original build number
[warn] ### Installed : 241.14494.277
[warn] ### Original  : 241.14494.81
[warn] Plugin [com.intellij.java] is already included by: [JUnit] in project 'scala-impl'
[warn] Plugin [com.intellij.properties] is already included by: [com.intellij.java-i18n] in project 'properties'
[warn] Plugin [com.intellij.java] is already included by: [JUnit] in project 'testing-support'
[warn] Plugin [com.intellij.java] is already included by: [JUnit] in project 'uast'
[warn] Plugin [com.intellij.java] is already included by: [org.jetbrains.bsp] in project 'intellij-bsp'
[warn] Plugin [com.intellij.properties] is already included by: [org.jetbrains.idea.maven, com.intellij.java-i18n] in project 'scalaCommunity'
[warn] Plugin [com.intellij.java] is already included by: [JUnit, org.jetbrains.bsp] in project 'scalaCommunit
```

I guess we'll have to exclude some plugins to avoid doubling them, but I have no idea how to do it with sbt. We have to consult the Scala team probably.

**Comment by agluszak** [comment]
Running the plugin doesn't work tho:

```
Executing pre-compile tasks…
Cleaning output directories…
Running 'before' tasks
Checking sources
Copying resources… [scalaCommunity.compiler-shared]
Copying resources… [scalaCommunity.decompiler]
Searching for compilable files... [scalaCommunity.scalatest-finders]
Searching for compilable files... [scalaCommunity.scala-utils-language-rt]
Searching for compilable files... [scalaCommunity.scala-utils-language]
Searching for compilable files... [scalaCommunity.test-runners]
Reading compilation settings... [scalaCommunity.scalatest-finders]
Reading compilation settings... [scalaCommunity.scala-utils-language-rt]
Reading compilation settings... [scalaCommunity.scala-utils-language]
Reading compilation settings... [scalaCommunity.test-runners]
scala: compiling 2 Scala sources and 1 Java source to /home/andrzej.gluszak/code/jetbrains/zpp/intellij-scala/scala/scala-utils-language-rt/target/scala-2.13/classes ...
scala: compiling 1 Scala source to /home/andrzej.gluszak/code/jetbrains/zpp/intellij-scala/scala/scala-utils-language/target/scala-2.13/classes ...
scala: compiling 26 Java sources to /home/andrzej.gluszak/code/jetbrains/zpp/intellij-scala/scala/test-integration/test-runners/target/scala-2.13/classes ...
scala: compiling 23 Java sources to /home/andrzej.gluszak/code/jetbrains/zpp/intellij-scala/scala/test-integration/scalatest-finders/target/classes ...
scala: source value 8 is obsolete and will be removed in a future release

scala: target value 8 is obsolete and will be removed in a future release

scala: To suppress warnings about obsolete options, use -Xlint:-options.

scala: done compiling
Adding nullability assertions… [scalaCommunity.scalatest-finders]
Adding threading assertions… [scalaCommunity.scalatest-finders]
Adding pattern assertions… [scalaCommunity.scalatest-finders]
Adding nullability assertions… [scalaCommunity.test-runners]
Searching for compilable files... [scalaCommunity.scala-api]
Reading compilation settings... [scalaCommunity.scala-api]
scala: compiling 8 Scala sources and 5 Java sources to /home/andrzej.gluszak/code/jetbrains/zpp/intellij-scala/scala/scala-api/target/scala-2.13/classes ...
Adding threading assertions… [scalaCommunity.test-runners]
Adding pattern assertions… [scalaCommunity.test-runners]
Copying resources… [scalaCommunity.sbt-api]
Copying resources… [scalaCommunity.runners]
Searching for compilable files... [scalaCommunity.runners]
Reading compilation settings... [scalaCommunity.runners]
scala: compiling 1 Java source to /home/andrzej.gluszak/code/jetbrains/zpp/intellij-scala/scala/runners/target/scala-2.13/classes ...
Adding nullability assertions… [scalaCommunity.scala-utils-language]
Adding threading assertions… [scalaCommunity.scala-utils-language]
Adding pattern assertions… [scalaCommunity.scala-utils-language]
Adding nullability assertions… [scalaCommunity.runners]
Adding threading assertions… [scalaCommunity.runners]
Searching for compilable files... [scalaCommunity.decompiler]
Reading compilation settings... [scalaCommunity.decompiler]
Adding pattern assertions… [scalaCommunity.runners]
scala: compiling 10 Scala sources to /home/andrzej.gluszak/code/jetbrains/zpp/intellij-scala/scala/decompiler/target/scala-2.13/classes ...
Copying resources… [scalaCommunity.scala-impl]
Copying resources… [scalaCommunity.sbt-impl]
Searching for compilable files... [tests of scalaCommunity.scalatest-finders]
Searching for compilable files... [tests of scalaCommunity.scala-utils-language]
Copying resources… [scalaCommunity.tasty-reader]
Searching for compilable files... [tests of scalaCommunity.test-runners]
Searching for compilable files... [tests of scalaCommunity.scala-api]
Searching for compilable files... [tests of scalaCommunity.sbt-api]
Searching for compilable files... [tests of scalaCommunity.runners]
Searching for compilable files... [tests of scalaCommunity.compiler-shared]
Copying resources… [scalaCommunity.worksheet-repl-interface]
Searching for compilable files... [scalaCommunity.worksheet-repl-interface]
Reading compilation settings... [scalaCommunity.worksheet-repl-interface]
scala: compiling 4 Java sources to /home/andrzej.gluszak/code/jetbrains/zpp/intellij-scala/scala/worksheet-repl-interface/target/classes ...
Adding nullability assertions… [scalaCommunity.worksheet-repl-interface]
Adding threading assertions… [scalaCommunity.worksheet-repl-interface]
Adding pattern assertions… [scalaCommunity.worksheet-repl-interface]
Searching for compilable files... [scalaCommunity.jps]
Reading compilation settings... [scalaCommunity.jps]
scala: compiling 3 Scala sources to /home/andrzej.gluszak/code/jetbrains/zpp/intellij-scala/scala/jps/target/scala-2.13/classes ...
Adding nullability assertions… [scalaCommunity.scala-utils-language-rt]
Adding threading assertions… [scalaCommunity.scala-utils-language-rt]
Adding pattern assertions… [scalaCommunity.scala-utils-language-rt]
Searching for compilable files... [scalaCommunity.compiler-shared]
Reading compilation settings... [scalaCommunity.compiler-shared]
scala: compiling 59 Scala sources and 8 Java sources to /home/andrzej.gluszak/code/jetbrains/zpp/intellij-scala/scala/compiler-shared/target/scala-2.13/classes ...
Adding nullability assertions… [scalaCommunity.jps]
Adding threading assertions… [scalaCommunity.jps]
Adding pattern assertions… [scalaCommunity.jps]
Searching for compilable files... [scalaCommunity.tasty-reader]
Reading compilation settings... [scalaCommunity.tasty-reader]
scala: compiling 10 Scala sources to /home/andrzej.gluszak/code/jetbrains/zpp/intellij-scala/scala/tasty-reader/target/scala-3.3.1/classes ...
Adding nullability assertions… [scalaCommunity.scala-api]
Adding threading assertions… [scalaCommunity.scala-api]
Adding pattern assertions… [scalaCommunity.scala-api]
Searching for compilable files... [tests of scalaCommunity.scala-utils-language-rt]
Reading compilation settings... [tests of scalaCommunity.scala-utils-language-rt]
scala: compiling 1 Scala source to /home/andrzej.gluszak/code/jetbrains/zpp/intellij-scala/scala/scala-utils-language-rt/target/scala-2.13/test-classes ...
Adding nullability assertions… [tests of scalaCommunity.scala-utils-language-rt]
Adding threading assertions… [tests of scalaCommunity.scala-utils-language-rt]
Adding pattern assertions… [tests of scalaCommunity.scala-utils-language-rt]
Searching for compilable files... [tests of scalaCommunity.workspace-entities]
Copying resources… [scalaCommunity.compiler-integration]
Copying resources… [scalaCommunity.structure-view]
Searching for compilable files... [tests of scalaCommunity.jps]
Copying resources… [scalaCommunity.testing-support]
Copying resources… [scalaCommunity.codeInsight]
Copying resources… [scalaCommunity.repl]
Copying resources… [scalaCommunity.intelliLang]
Copying resources… [scalaCommunity.worksheet]
Copying resources… [scalaCommunity.uast]
Copying resources… [scalaCommunity.properties]
Copying resources… [scalaCommunity.ml-completion]
Copying resources… [scalaCommunity.maven]
Copying resources… [scalaCommunity.conversion]
Copying resources… [scalaCommunity.debugger]
Copying resources… [scalaCommunity.devKit]
Copying resources… [scalaCommunity.intellij-bsp]
Copying resources… [scalaCommunity.gradle]
Copying resources… [scalaCommunity.pluginXml]
Searching for compilable files... [tests of scalaCommunity.worksheet-repl-interface]
Searching for compilable files... [tests of scalaCommunity.repl]
Searching for compilable files... [scalaCommunity.pluginXml]
Searching for compilable files... [scalaCommunity]
Searching for compilable files... [tests of scalaCommunity]
Searching for compilable files... [tests of scalaCommunity.pluginXml]
Adding nullability assertions… [scalaCommunity.decompiler]
Adding threading assertions… [scalaCommunity.decompiler]
Adding pattern assertions… [scalaCommunity.decompiler]
Searching for compilable files... [tests of scalaCommunity.decompiler]
Reading compilation settings... [tests of scalaCommunity.decompiler]
scala: compiling 2 Scala sources to /home/andrzej.gluszak/code/jetbrains/zpp/intellij-scala/scala/decompiler/target/scala-2.13/test-classes ...
Adding nullability assertions… [tests of scalaCommunity.decompiler]
Adding threading assertions… [tests of scalaCommunity.decompiler]
Adding pattern assertions… [tests of scalaCommunity.decompiler]
Adding nullability assertions… [scalaCommunity.compiler-shared]
Adding threading assertions… [scalaCommunity.compiler-shared]
Adding pattern assertions… [scalaCommunity.compiler-shared]
Searching for compilable files... [scalaCommunity.sbt-api]
Reading compilation settings... [scalaCommunity.sbt-api]
scala: compiling 21 Scala sources and 5 Java sources to /home/andrzej.gluszak/code/jetbrains/zpp/intellij-scala/sbt/sbt-api/target/scala-2.13/classes ...
scala: Missing required plugin: scalac-patches

Errors occurred while compiling module 'scalaCommunity.sbt-api'
Adding nullability assertions… [scalaCommunity.tasty-reader]
Adding threading assertions… [scalaCommunity.tasty-reader]
Adding pattern assertions… [scalaCommunity.tasty-reader]
Finished, saving caches…
Builder Scala sbt builder requested build stop
Executing post-compile tasks…
Synchronizing output directories…
04/04/2024, 14:26 - Build completed with 1 error and 3 warnings in 38 sec, 910 ms
```

**Comment by AnthonyGrod** [comment]
@agluszak isn't the error you've encountered a good old "missing scalac-patches" error that should be resolved by rebuilding the project? It doesn't appear on my machine

**Comment by p-fuchs** [comment]
@agluszak I'm cleaning the branches, so I'm going to merge this PR.
@AnthonyGrod will address other points if needed, but we'd like to test it with other improvements which we've made on the main branch.

**Pull request merged**